numpy/tests/typing/fail/fromnumeric.py

```diff
@@ -124,3 +124,31 @@
 np.amin(a, out=1.0)  # E: No overload variant of "amin" matches argument type
 np.amin(a, initial=[1.0])  # E: No overload variant of "amin" matches argument type
 np.amin(a, where=[1.0])  # E: List item 0 has incompatible type
+
+np.prod(a, axis=1.0)  # E: No overload variant of "prod" matches argument type
+np.prod(a, out=False)  # E: No overload variant of "prod" matches argument type
+np.prod(a, keepdims=1.0)  # E: No overload variant of "prod" matches argument type
+np.prod(a, initial=int)  # E: No overload variant of "prod" matches argument type
```

Comment:
This is legal for suitable a:
```
a = np.array([], dtype=object)
np.prod(a, initial=int)
```

Reply:
Incidentally, this also proves that the return type of `np.prod` can be anything, not just `number`.

Reply:
> Incidentally, this also proves that the return type of `np.prod` can be anything, not just `number`.

If I'm not mistaken this is limited to, the notoriously difficult to type, object arrays.
If so I feel it'd be better to leave things as they are for now, especially since this appears to be a very specific (seemingly undocumented?) situation.

The alternative would be to set `number` and the return type to `Any`, 
which is practically useless from a typing standpoint.

Reply:
> especially since this appears to be a very specific (seemingly undocumented?) situation.

How so? `initial` is documented as a "scalar", not as a number.

Perhaps the compromise is to stick `# TODO: This is actually legal` or similar by each test case that currently checks for an error but should not.

Reply:
> How so? `initial` is documented as a "scalar", not as a number.

That's fair actually a good point; I more or less conflated scalar with numerical scalar here.

> Perhaps the compromise is to stick # TODO: This is actually legal or similar by each test case that currently expects and results in an error, but should not.

I agree, I'll push an update in a bit.

As for the future: 

What I've gathered is, in the case of object arrays, that `np.prod()` simply falls back to the `.__mult__()` method of the passed arrays individual elements and hence all that's required of `initial` is that its type is compatible with aforementioned elements. 
For object arrays we may or may not be able to easily express this with a `__mult__` Protocol once ndarray is generic over its dtype, though this will depend on the exact implementation details.

Reply:
> Perhaps the compromise is to stick `# TODO: This is actually legal` or similar by each test case that currently expects and results in an error, but should not.

Added in https://github.com/numpy/numpy/pull/16729/commits/3bc51b8f1c3af1dbf17d6b84084ce6b914b97c52.

Reply:
> For object arrays we may or may not be able to easily express this with a `__mult__` Protocol once ndarray is generic over its dtype, though this will depend on the exact implementation details.

I wouldn't personally bother. Simply having `prod(ndarray[int], initial: int) -> int`,  `prod(ndarray[object], initial: object) -> object` etc is all that's really needed. The details of exactly what the object loops of ufuncs do and do not support is nort really interesting.